cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/kafka/KafkaLogAppender.java

```diff
@@ -50,13 +60,27 @@
   @Override
   protected void append(LogMessage logMessage) {
     try {
+      String partitionKey = getPartitionKey(logMessage.getLoggingContext());
       byte [] bytes = loggingEventSerializer.toBytes(logMessage.getLoggingEvent(), logMessage.getLoggingContext());
-      producer.publish(logMessage.getLoggingContext().getLogPartition(), bytes);
+      producer.publish(partitionKey, bytes);
     } catch (Throwable t) {
       LOG.error("Got exception while serializing log event {}.", logMessage.getLoggingEvent(), t);
     }
   }
 
+  private String getPartitionKey(LoggingContext loggingContext) {
+    String namespaceId = loggingContext.getSystemTagsMap().get(NamespaceLoggingContext.TAG_NAMESPACE_ID).getValue();
+
+    if (cConf.get(LoggingConfiguration.KAFKA_PARTITION_KEY).equals(KAFKA_PARTITION_KEY_PROGRAM) ||
+      namespaceId.equals(NamespaceId.SYSTEM.getEntityName())) {
+      return loggingContext.getLogPartition();
```

Comment:
What do you mean explicitly? We do not get MDCPropertyMap at publisher side. Also I think when we remove LoggingContext#getLogPartition(), we should remove all the usages of it.